compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java

```diff
@@ -578,6 +578,11 @@ private static void populateImmutableStructureFields(Types types, SymbolTable sy
                 invokableSymbol.restParam = tsymbol.restParam;
                 invokableSymbol.retType = tsymbol.returnType;
                 invokableSymbol.flags = tsymbol.flags;
+            } else if (immutableFieldType == symTable.semanticError) {
```

Comment:
why we need to check `immutableFieldType == symTable.semanticError` here

Reply:
`immutableFieldType` will be `symTable.semanticError` when the field type can never be readonly. This can only happen for optional fields, so we create an optional field of type `never` for it. 

If the field wasn't optional, we wouldn't get here.

For example, where `Foo` is

```ballerina
type Foo record {
    int a;
    stream<int> b?;
};
```

when we do `Foo & readonly`, we get here for `b`.

`Foo & readonly` is effectively

```ballerina
readonly & record {
    int a;
    never b?;
}
```